# Web Scraping and time series

In [1]:
#import the necessary packages
import requests
import pandas as pd
from selenium import webdriver
from selenium.webdriver.support.ui import Select
from urllib.request import Request, urlopen
from selenium.webdriver.common.keys import Keys
import regex as re
import matplotlib as plt
from bs4 import BeautifulSoup as bsoup
%matplotlib inline
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

### go to the Eurosport Paris Roubaix cycling results page and load it into a pandas dataframe

https://www.eurosport.com/cycling/paris-roubaix/2022/live-road-race-men_mtc1323920/live.shtml

Hint, pandas read_html() will fail to find the table even though it is on the page.  You need to find the table (class='standing-table') with a selenium driver.get() or BeautifulSoup, then pass that to pandas read_html().

In [8]:
# Reading in the web page Then displaying

# Eurosport Paris Roabix cycling results
url = "https://www.eurosport.com/cycling/paris-roubaix/2022/live-road-race-men_mtc1323920/live.shtml"

# Setting options to not open the browser
options = webdriver.ChromeOptions()
options.add_argument("headless")

# Placed the Driver in the HW File
DRIVER_PATH = 'chromedriver'
driver = webdriver.Chrome(executable_path=DRIVER_PATH, options=options)

# Labeling and getting the Website
driver.get(url)

In [13]:
# Passing the webpage to Beautiful Soup to parse through
soup = bsoup(driver.page_source, 'html.parser')
table = soup.find('table', {'class':'standing-table'})
rows = table.find_all('tr', {'class':'standing-table__row'})
for row in rows:
    print(row)
    print("\nNew Line")
# df_table = pd.read_html(table)
# df_table


<tr class="standing-table__row"><td class="standing-table__cell standing-table__cell--position">1</td><td class="standing-table__cell standing-table__cell--player"><a class="standing-table__player-link" href="/cycling/dylan-van-baarle_prs271291/person.shtml"><span class="player"><img alt="" class="isg-interchange lazyload-ok" data-isg-interchange='{"f":{"medium":"https://i.eurosport.com/_iss_/person/pp_clubteam/medium/763405.jpg"},"b":{"ms":"medium","ml":"medium","t":"medium","d":"medium"}}' data-isg-lazy="https://i.eurosport.com/_iss_/person/pp_clubteam/medium/763405.jpg" data-lazy-type="isg" src="https://i.eurosport.com/_iss_/person/pp_clubteam/medium/763405.jpg" title="Dylan van Baarle"/><span class="country"><img alt="" class="isg-interchange lazyload-ok" data-isg-lazy="https://i.eurosport.com/_iss_/geo/country/flag/medium/2201.png" data-lazy-type="isg" src="https://i.eurosport.com/_iss_/geo/country/flag/medium/2201.png" title="Netherlands"/></span></span><span class="text"><span c

### For the next part of the homework you will be scraping data from the following URL

# US FDA web page
https://www.accessdata.fda.gov/scripts/cder/daf/

You are going to go to the FDA web page and search for Fentanyl.  It should return a list of all drugs that contain fentanyl.  The table that comes back is just a simple list of the drug names.  If you click on the drug name, it expands the row to show Drug Name, Link, NDA number, Dosage Form, Marketing Status, and Manufacturer.  You need to scrape this web page and print those values seperated by commas.

### POST forms
POST forms do not put the data in the URL, the data is submitted via javascript.  You have to pretend to fill out the form and submit it.  Grabbing the parameters
First we need to find out what parameters we’re going to hunt down. To do this, first make your way to the form, then get prepared.

1) In Chrome, more tools > Developer Tools 2) Click the Network tab 3) Fill the form out, and submit it 4) Scroll up to the top of the Network pane, select the segment of the URL you’re at (I’m at tempai.cfm) 5) Click it 6) Select Headers on the right 7) Scroll down until you see Form Data

### 1)  Part 1, enter the search term fentanyl and submit

In [187]:


from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By

# Placed the Driver in the HW File
DRIVER_PATH = 'chromedriver'
driver = webdriver.Chrome(executable_path=DRIVER_PATH)

# Labeling and getting the Website
url = 'https://www.accessdata.fda.gov/scripts/cder/daf/'
driver.get(url)

# Quering for Fentanyl and submitting
driver.find_element(By.NAME, "searchterm").send_keys("fentanyl" + Keys.ENTER)


### 2) Pull the data from the website, check the status code to make sure it was successfull and create a BeautifulSoup object and parse the data.  Print the soup object below:

In [272]:
pageSource = driver.page_source
soup = bsoup(pageSource, 'html.parser')

df = pd.DataFrame(columns=["Drug Name", "Link", "NDA number", "Dosage Form", "Marketing Status", "Manufacturer"])
# Grabbing all the Results
results = soup.find_all('tr')


for result in results:
    # Grabbing the links from the webpage
    inner_link = result.find_all('a', href=True)
    for a in inner_link:
        if len(a['href']) > 11:
            row_link = str("https://www.accessdata.fda.gov" + a['href'])
    
    # Grabbing the table results
    inner = result.find_all('li')
    inner_text = inner[0].get_text()
    panda_row  = inner_text.split('|')

    # Placing results in to panda
    df = df.append({"Drug Name": str(panda_row[0])
                    , "Link": row_link
                    , "NDA number": str(panda_row[1])
                    , "Dosage Form": str(panda_row[2])
                    , "Marketing Status": str(panda_row[3])
                    , "Manufacturer": str(panda_row[4]).strip()}, ignore_index=True)

df

Drug Name  \
0                         ABSTRAL (FENTANYL CITRATE)    
1                           ACTIQ (FENTANYL CITRATE)    
2                           DURAGESIC-100 (FENTANYL)    
3                            DURAGESIC-12 (FENTANYL)    
4                            DURAGESIC-25 (FENTANYL)    
5                            DURAGESIC-37 (FENTANYL)    
6                            DURAGESIC-50 (FENTANYL)    
7                            DURAGESIC-75 (FENTANYL)    
8                        FENTANYL (FENTANYL CITRATE)    
9                FENTANYL CITRATE (FENTANYL CITRATE)    
10  FENTANYL CITRATE AND DROPERIDOL (DROPERIDOL; F...   
11  FENTANYL CITRATE PRESERVATIVE FREE (FENTANYL C...   
12                           FENTANYL-100 (FENTANYL)    
13                            FENTANYL-12 (FENTANYL)    
14                            FENTANYL-25 (FENTANYL)    
15                            FENTANYL-37 (FENTANYL)    
16                            FENTANYL-50 (FENTANYL)    
17                            FENTANYL-62 (FENTANYL)    
18                            FENTANYL-75 (FENTANYL)    
19                            FENTANYL-87 (FENTANYL)    
20                        FENTORA (FENTANYL CITRATE)    
21            INNOVAR (DROPERIDOL; FENTANYL CITRATE)    
22                   IONSYS (FENTANYL HYDROCHLORIDE)    
23                        LAZANDA (FENTANYL CITRATE)    
24                        ONSOLIS (FENTANYL CITRATE)    
25    SUBLIMAZE PRESERVATIVE FREE (FENTANYL CITRATE)    
26                                 SUBSYS (FENTANYL)    

                                                 Link       NDA number  \
0   https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #022510    
1   https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #020747    
2   https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #019813    
3   https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #019813    
4   https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #019813    
5   https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #019813    
6   https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #019813    
7   https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #019813    
8   https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #020195    
9   https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #019115    
10  https://www.accessdata.fda.gov/scripts/cder/da...   ANDA  #071982    
11  https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #019101    
12  https://www.accessdata.fda.gov/scripts/cder/da...   ANDA  #076258    
13  https://www.accessdata.fda.gov/scripts/cder/da...   ANDA  #076258    
14  https://www.accessdata.fda.gov/scripts/cder/da...   ANDA  #076258    
15  https://www.accessdata.fda.gov/scripts/cder/da...   ANDA  #076258    
16  https://www.accessdata.fda.gov/scripts/cder/da...   ANDA  #076258    
17  https://www.accessdata.fda.gov/scripts/cder/da...   ANDA  #076258    
18  https://www.accessdata.fda.gov/scripts/cder/da...   ANDA  #076258    
19  https://www.accessdata.fda.gov/scripts/cder/da...   ANDA  #076258    
20  https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #021947    
21  https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #016049    
22  https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #021338    
23  https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #022569    
24  https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #022266    
25  https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #016619    
26  https://www.accessdata.fda.gov/scripts/cder/da...   NDA   #202788    

                             Dosage Form Marketing Status         Manufacturer  
0                     TABLET;SUBLINGUAL     Discontinued   SENTYNL THERAPS INC  
1           TROCHE/LOZENGE;TRANSMUCOSAL     Prescription              CEPHALON  
2    FILM, EXTENDED RELEASE;TRANSDERMAL     Discontinued        JANSSEN PHARMS  
3    FILM, EXTENDED RELEASE;TRANSDERMAL     Discon

### 3) import selenium webdriver and BeautifulSoup

Get your webdriver and get the page.  Submit the search term and click the search button.

In [274]:
# I believe I have done this above
df['Link'].to_list()

['https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=022510',
 'https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=020747',
 'https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=019813',
 'https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=019813',
 'https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=019813',
 'https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=019813',
 'https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=019813',
 'https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=019813',
 'https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=202577',
 'https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=212086',
 'https://

### 4) run the selenium page_source through beautiful soup using the html parser

In [182]:
# About to go through the page
# Grabbing all the Results


### 5) Grab all the rows from the table using find_all.  Note that some of the data is in a 'tr' header and some is in 'li'.

Parse the data and print the results

Note the link on the page is a relative link.  You must join that link with 'https://www.accessdata.fda.gov' to make it a link you can click on.
Example:  https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=022510

### The output should look like this (except print all rows)
Drug Name,Link,NDA number,Dosage Form,Marketing Status,Manufacturer

ABSTRAL , https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=022510 , NDA   #022510 , TABLET;SUBLINGUAL , Discontinued , SENTYNL THERAPS INC

ACTIQ , https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=020747 , NDA   #020747 , TROCHE/LOZENGE;TRANSMUCOSAL , Prescription , CEPHALON

DURAGESIC-100 , https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=019813 , NDA   #019813 , FILM, EXTENDED RELEASE;TRANSDERMAL , Discontinued , JANSSEN PHARMS

DURAGESIC-12 , https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=019813 , NDA   #019813 , FILM, EXTENDED RELEASE;TRANSDERMAL , Discontinued , JANSSEN PHARMS

DURAGESIC-25 , https://www.accessdata.fda.gov/scripts/cder/daf/index.cfm?event=overview.process&ApplNo=019813 , NDA   #019813 , FILM, EXTENDED RELEASE;TRANSDERMAL , Discontinued , JANSSEN PHARMS


### 10 bonus points for following the link and getting the strength information from that page.  Print the strength information out along with everything you were getting above.